### PR TITLE
feat: load local config

### PR DESCRIPTION
**Pull request by AlexanderBrevig**
This will enable having a `.helix/config.toml` in a project repo and have it override default and `~/.config/helix/config.toml` settings.

This behaviour is controlled with `config.toml`:
```toml
[editor.security]
load-local-config = false
confirm-local-config = true
```

The default keeps Helix working as it is. 
If you only set `load-local-config = true` then you'll be promted to answer `yes` before loading. 
This prompt can be removed by setting `confirm-local-config = false`.

### CODE REVIEW

File: helix-loader/src/main.rs [hunk 1]
```diff
@@ -5,5 +5,6 @@ use helix_loader::grammar::fetch_grammars;
 // compilation time. This is not meant to be run manually.
 
 fn main() -> Result<()> {
+    println!("uo");
```
Comment (kirawi) [inline helix-loader/src/main.rs:8]: You should delete this.

File: helix-term/src/main.rs [hunk 2]
```diff
@@ -114,24 +116,40 @@ FLAGS:
...
-                let _ = std::io::stdin().read(&mut []);
-                Config::default()
-            }),
-        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Config::default(),
-        Err(err) => return Err(Error::new(err)),
+    let config_value = local_config_dirs()
+        .into_iter()
+        .chain([config_dir()].into_iter())
+        .map(|path| path.join("config.toml"))
+        .rev()
```
Comment (kirawi) [inline helix-term/src/main.rs:123]: Why call `rev()`? I think that would make the local configs be overridden by the upper configs.

File: helix-loader/src/main.rs (hunk 1 above)
Comment (AlexanderBrevig) [reply helix-loader/src/main.rs:8]: Wow. I have no idea how or why that's there. I'll go through and fix lint issues and of course remove this. Thanks!

**Comment by AlexanderBrevig** [comment]
Thanks for the review and input @kirawi! I've fixed the requested changes (an embarrassing reminder that late night sessions may lead to weird things...) and fixed clippy lint errors locally.

**Review by AlexanderBrevig — COMMENTED** [review]
Maybe it should use `eprintln` and inform the user on startup (file and error message) much like the current behaviour.

Probably also should update the docs. Will fix this tomorrow (no computer today).

:heavy_check_mark: This is now implemented.

**Comment by d12bb** [comment]
This should be optional, as one may not want to use custom config coming with someone else's repo and, more important, this could harm the user's computer, for example by resetting a common key bind to some harmful shell command.

**Comment by the-mikedavis** [comment]
That probably falls under something like https://github.com/helix-editor/helix/issues/2697 (or is similar in spirit). It's notable that we already merge in `.helix/languages.toml` and I think you could easily craft a malicious `languages.toml` like so:

```toml
[[language]]
name = "toml"
language-server = { command = "rm", args = ["-rf", "/"] }
```

**Comment by AlexanderBrevig** [comment]
I've added this to `config.toml`

```toml
[editor.security]
load-local-config = false
confirm-local-config = true
```

(commented out, you will also see `-languages` versions for a future PR)

That's the default. If you toggle both, then Helix will load `$PWD/.helix/config.toml` without asking to do so.

What do you think?
I very much like the feature, but I am not married to the implementation details. I'll fix anything you see fit.

**Comment by AlexanderBrevig** [comment]
Updated the docs and merged with master. `hx -c ~/team/hx/config.toml` will now load the "team" config, and then ask to load `$PWD/.helix/config.toml` if `[editor.security] load-local-config = true`.

Anything I can do to help this PR forward? (I understand the team is under a lot of load)

**Comment by AlexanderBrevig** [comment]
I've fixed your suggestions @kirawi, do you have some more feedback or questions? :)

File: helix-term/src/main.rs [hunk 3]
```diff
@@ -115,25 +115,72 @@ FLAGS:
...
-            }),
-        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Config::default(),
-        Err(err) => return Err(Error::new(err)),
+    // We need the config, to get the editor.security.[load|confirm]_local_config flags
+    let root_config = match std::fs::read_to_string(helix_loader::config_file())
+        .map(|config| toml::from_str(&config))?
+    {
+        Ok(config) => config,
+        Err(_) => Config::default(),
+    };
```
Comment (pickfire) [inline helix-term/src/main.rs:126]:
```suggestion
    let root_config = match std::fs::read_to_string(helix_loader::config_file())
        .map(|config| toml::from_str(&config))?
        .unwrap_or_default();
```
I think this should work.

Comment (AlexanderBrevig) [reply helix-term/src/main.rs:126]: I will try

File: helix-term/src/main.rs [hunk 4]
```diff
@@ -115,25 +115,69 @@ FLAGS:
...
+            .try_into()?,
+        Err(err) => {
+            eprintln!("Bad configs: {:#?}", err);
+            eprintln!("Press <ENTER> to continue with default config");
+            use std::io::Read;
+            // This waits for an enter press.
+            let _ = std::io::stdin().read(&mut []);
+            Config::default()
+        }
     };
```
Comment (pickfire) [inline helix-term/src/main.rs:181]: Please move these into a function, it became quite huge.

File: helix-term/src/main.rs [hunk 5]
```diff
@@ -115,25 +115,69 @@ FLAGS:
...
-                eprintln!("Bad config: {}", err);
-                eprintln!("Press <ENTER> to continue with default config");
-                use std::io::Read;
-                let _ = std::io::stdin().read(&mut []);
-                Config::default()
-            }),
-        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Config::default(),
-        Err(err) => return Err(Error::new(err)),
+    // We need the config, to get the editor.security.[load|confirm]_local_config flags
+    let root_config = std::fs::read_to_string(helix_loader::config_file())
```
Comment (pickfire) [inline helix-term/src/main.rs:121]:
```suggestion
    let root_config: Config = std::fs::read_to_string(helix_loader::config_file())
```

File: helix-term/src/main.rs [hunk 6]
```diff
@@ -115,25 +115,69 @@ FLAGS:
...
+        .map(|config| toml::from_str(&config))?
+        .unwrap_or_default();
+
+    // Now, let's load config based on a PathBuf
+    // We will keep it toml::Value so we can use helix_loader::merge_toml_values
+    // This will respect editor.security settings
+    let load_config = |config_path: std::path::PathBuf| {
+        if std::path::Path::new(&config_path).exists() {
```
Comment (pickfire) [inline helix-term/src/main.rs:129]:
```suggestion
        if config_path.exists() {
```
This auto deref to `Path`, so don't need to manually create new `Path`.

File: helix-term/src/main.rs [hunk 7]
```diff
@@ -115,25 +115,69 @@ FLAGS:
...
+                    if root_config.editor.security.confirm_local_config {
+                        eprintln!(
+                            "Type yes to continue with loading config: {:#?}",
+                            config_path
+                        );
+                        use std::io::Read;
+                        confirmed = match std::io::stdin().read(&mut [1, 2, 3, 4]) {
+                            Ok(n) => n == "yesE".len(),
+                            Err(_) => false,
+                        }
```
Comment (pickfire) [inline helix-term/src/main.rs:142]: What is this doing?

Comment (AlexanderBrevig) [reply helix-term/src/main.rs:142]: That whole thing is embarrasing. It should rather match on the actual input of `yes`. I thought I had fixed that to be honest. I'll make a pass on it now and refactor out that function you requested.

**Comment by AlexanderBrevig** [comment]
Huge thank you @pickfire I can see that I maybe wrote this code a bit prematurely in terms of how well (let's be honest, bad) I knew Rust.

However, the strange thing now (before the refactor too) is that I seem to lose ability to input text after the editor is loaded. It did not use to be that way.

Reproduce:
Add `[editor.security] load-local-config = true`  and maybe some config file in `./helix/config.toml`.
Then `cargo run .` will first ask you to type `yes<RET>` if you have a file, then open the file picker - here you can type. When you open a file, suddenly I cannot type any more.
I have very little idea where to check, and somehow doubt that its this code that introduces this error. Unless `std::io::stdin().read_line` somehow blocks the events coming in after...?

I did some cleanup in addition to yours. It is better now, at least. A bit unsire if the functions should be in config.rs or simply at the bottom of main.rs.

File: helix-term/src/main.rs (hunk 4 above)
Comment (AlexanderBrevig) [reply helix-term/src/main.rs]: I have refactored the entire thing. Let me know if I should do another pass at the Rust, but to me it is a lot better now. Thanks again for your time and input!

(I don't know what will generate notifications but I will presume my comment and the "resolve conversation" will suffice).

File: book/src/configuration.md [hunk 8]
```diff
@@ -229,3 +233,13 @@ Example:
...
+
+### `[editor.security]` Section
+
+Opt in to features that may put you at risk. 
+It is possible to write malicious TOML files, so we suggest you keep the `confirm_*` options to their default value of `true`.
+
+| Key                    | Description                                                                             | Default |
+| ---                    | ---                                                                                     | ---     |
+| `load_local_config`    | Load a `config.yaml` from `$PWD/.helix` that will merge with your global configuration. | `false` |
+| `confirm_local_config` | Prompt the user at launch in order to accept the loading of a found local `$PWD/.helix` | `true`  |
```
Comment (pickfire) [inline book/src/configuration.md:245]: Not sure if this is just me but I don't find this quite important.

File: helix-term/src/config.rs [hunk 9]
```diff
@@ -55,6 +55,95 @@ impl Config {
...
+            .into_iter()
+            .map(|path| path.join("config.toml"))
+            .chain([helix_loader::config_file()])
+            .filter_map(|file| Config::load_config_toml_values(&root_config, file));
+
+        // Merge configs and return, or alert user of error and load default
+        match local_config_values
+            .fold(toml::Value::Table(toml::value::Table::default()), |a, b| {
+                helix_loader::merge_toml_values(b, a, 3)
+            })
```
Comment (pickfire) [inline helix-term/src/config.rs:81]: I believe we can use `reduce` instead.

File: helix-term/src/config.rs [hunk 10]
```diff
@@ -55,6 +55,95 @@ impl Config {
...
+            let mut confirmed = true;
+            if config_path != helix_loader::config_file() {
+                if root_config.editor.security.load_local_config {
+                    if root_config.editor.security.confirm_local_config {
+                        eprintln!(
+                            "Type yes<ENTER> to continue with loading config: {:#?}",
+                            config_path
+                        );
+                        let mut input = String::new();
+                        std::io::stdin().read_line(&mut input).unwrap_or_default();
```
Comment (pickfire) [inline helix-term/src/config.rs:105]: >  Then cargo run . will first ask you to type yes<RET> if you have a file, then open the file picker - here you can type. When you open a file, suddenly I cannot type any more.

Most likely what you mentioned is due to this line, `read_line` probably interfere with raw input mode or something, I am not sure.

File: helix-term/src/config.rs [hunk 11]
```diff
@@ -55,6 +55,95 @@ impl Config {
...
+        }
+    }
+
+    // Load a specific config file if allowed by config
+    // Stay with toml::Values as they can be merged
+    pub fn load_config_toml_values(
+        root_config: &Config,
+        config_path: std::path::PathBuf,
+    ) -> Option<toml::Value> {
+        if config_path.exists() {
```
Comment (pickfire) [inline helix-term/src/config.rs:95]: Doing fast exit on huge nested function is good, like
```rust
        if !config_path.exists() {
            return None;
        }

File: helix-term/src/config.rs (hunk 10 above)
Comment (AlexanderBrevig) [reply helix-term/src/config.rs:105]: It turned out to be some setting in my config.toml
I should narrow it down (but I suspect it was my escape rebind)

**Review by pickfire — COMMENTED** [review]
`load_local_config` part seemed okay to me but I am not sure about the `confirm_local_config` part.

**Comment by AlexanderBrevig** [comment]
> `load_local_config` part seemed okay to me but I am not sure about the `confirm_local_config` part.

I am in favor of removing the confirmation prompt. Should I do it?

**Comment by pickfire** [comment]
> I am in favor of removing the confirmation prompt. Should I do it?

Yes, I think for now we can keep it with a small change to only have a config to allow loading local config. That would be easier review and try it out.

File: book/src/configuration.md [hunk 12]
```diff
@@ -229,3 +233,12 @@ Example:
 render = true
 character = "╎"
 ```
+
+### `[editor.security]` Section
+
+Opt in to features that may put you at risk. 
+It is possible to write malicious TOML files, so we suggest you keep the `confirm_*` options to their default value of `true`.
```
Comment (pickfire) [inline book/src/configuration.md:240]: I think the confirm part is outdated.

File: helix-term/src/config.rs [hunk 13]
```diff
@@ -87,11 +87,9 @@ impl Config {
...
         config_path: std::path::PathBuf,
     ) -> Option<toml::Value> {
-        if !config_path.exists() {
-            return None;
-        }
-        if config_path == helix_loader::config_file()
-            || root_config.editor.security.load_local_config
+        if config_path.exists()
+            && (config_path == helix_loader::config_file()
+                || root_config.editor.security.load_local_config)
```
Comment (pickfire) [inline helix-term/src/config.rs:92]: Still same thing, this 3 can just `return None`; to reduce one indentation level.

Comment (AlexanderBrevig) [reply helix-term/src/config.rs]: ...embarrasing, thanks!

**Comment by the-mikedavis** [comment]
The security aspect of this still bothers me a bit - it ends up really nerfing the feature. I think it's silly in the same way that git's [cve-2022-24765](https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765) (the safe.directory one) is silly, and to my knowledge no other terminal based editor is careful about this. I would argue at least that this should be enabled by default.

**Comment by AlexanderBrevig** [comment]
> The security aspect of this still bothers me a bit - it ends up really nerfing the feature. I think it's silly in the same way that git's [cve-2022-24765](https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765) (the safe.directory one) is silly, and to my knowledge no other terminal based editor is careful about this. I would argue at least that this should be enabled by default.

Since three of four voices in this PR seems to agree, I'll change the default.

**Review by pickfire — APPROVED** [review]
Looks good to me. Thanks for working on this.

File: book/src/configuration.md [hunk 14]
```diff
@@ -229,3 +233,11 @@ Example:
...
 character = "╎"
 ```
+
+### `[editor.security]` Section
+
+Opt in to features that may put you at risk. 
+
+| Key                    | Description                                                                             | Default |
+| ---                    | ---                                                                                     | ---     |
+| `load-local-config`    | Load `config.yaml` from `$PWD/.helix` that will merge with your global configuration.   | `false` |
```
[the remark this answers is not in the document]
Comment (AlexanderBrevig) [reply book/src/configuration.md]: Absolutely! Thanks!

File: book/src/configuration.md [hunk 15]
```diff
@@ -26,7 +26,11 @@ hidden = false
 ```
 
 You may also specify a file to use for configuration with the `-c` or
-`--config` CLI argument: `hx -c path/to/custom-config.toml`.
+`--config` CLI argument: `hx -c path/to/custom-config.toml`. 
+
+Finally, you can have a `config.toml` local to a project by it under a `.helix` directory in your repository.
```
[the remark this answers is not in the document]
Comment (AlexanderBrevig) [reply book/src/configuration.md:31]: Thanks! Will fix tonight 👍

**Comment by AlexanderBrevig** [comment]
Somehow I totally forgot this. Sorry.

**Comment by Nazariglez** [comment]
Hello fellows! This feature looks great. What needs to be done to be able to merge this? Thanks

**Comment by gibbz00** [comment]
> Hello fellows! This feature looks great. What needs to be done to be able to merge this? Thanks

I did a refactor on config loading, https://github.com/helix-editor/helix/pull/5636. I'm in the process of merging this PR with that branch, hoping that it can help getting this one merged too.

**Comment by Nazariglez** [comment]
That's awesome @gibbz00, thank you!

**Comment by pascalkuthe** [comment]
> I Hope I'm not too late, but I would like helix to also support `(config|languages).local.toml`, that way I can easily have a global gitignore for it and avoid polluting repos with my configs and/or ignores.
> 
> The exact behavior can either be local replaces the non-local file or is merged with it, favoring the local one (I prefer the second option but both can be argued for)

What exactly is the advantage of `config.local.toml` over `.helix/config.toml` in that regard?
I think adding `.helix/**` to the global gitignore would be enough (that would also take care of `.helix/languages.toml`)

**Comment by poliorcetics** [comment]
As helix becomes more used, I expect some repos will start having a config in their git, like many already do for VSCode, jetbrains IDEs or vim/neovim, in this case `.helix/**` won't be enough

**Comment by the-mikedavis** [comment]
`.helix/languages.toml` and `.helix/config.toml` also work in the global gitignore. I also don't see the advantage to using `config.local.toml` and `languages.local.toml`: it should be equivalent to `.helix/config.toml` and `.helix/languages.toml` respectively, no?

**Comment by poliorcetics** [comment]
No, what I mean is, in the project repo, people will start committing `<my repo>/.helix/{config,languages}.toml` (and ideally it should be `.config/helix/...` even locally, like `cargo-nextest` does), and having the hability to override it locally without having to play with the Git index would be nice.

As an example, search [`path:/.vscode/settings.json`](https://cs.github.com/?scopeName=All+repos&scope=&q=path%3A.vscode%2Fsettings.json). The results are murkier for Vim and Jetbrains IDEs but there are examples.

**Comment by the-mikedavis** [comment]
Oh I see, `config.local.toml` would be in addition to `.helix/config.toml`, right?

For scenarios like this I think it would be better to make use of the `-c <config-file>` CLI option and have a wrapper around `hx -c` to control which config is used based on the directory it's called in

**Comment by poliorcetics** [comment]
> For scenarios like this I think it would be better to make use of the `-c <config-file>` CLI option and have a wrapper around `hx -c` to control which config is used based on the directory it's called in

That seems like a very fast path to the disaster that is using `$CARGO_TARGET_DIR` globally and doesn't allow for merging configs in the local -> project -> user order, because projects may be settings configs to help users go in more smoothly and having to pass `-c` is simply a workaround

**Comment by charles-dyfis-net** [comment]
> The security aspect of this still bothers me a bit - it ends up really nerfing the feature. I think it's silly in the same way that git's [cve-2022-24765](https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765) (the safe.directory one) is silly, and to my knowledge no other terminal based editor is careful about this. I would argue at least that this should be enabled by default.

Both vim and emacs disallow modelines from configuring settings that can result in executables being run by default / without user approval.

**Pull request closed without merging**